lib/client.js

```diff
@@ -50,1444 +56,1613 @@ const socketErrors = [
+				let packet = null
+
+				if (!entry) {
+					this._resubscribe()
+					return
+				}
+
+				packet = entry.packet
+				this.log('deliver :: call _sendPacket for %o', packet)
+				let send = true
+				if (packet.messageId && packet.messageId !== 0) {
+					if (!this.messageIdProvider.register(packet.messageId)) {
+						send = false
+					}
+				}
+				if (send) {
+					this._sendPacket(packet, (err) => {
+						if (entry.cb) {
+							entry.cb(err)
+						}
+						deliver()
+					})
+				} else {
+					this.log(
+						'messageId: %d has already used. The message is skipped and removed.',
+						packet.messageId,
+					)
+					deliver()
+				}
+			}
+
+			this.log('connect :: sending queued packets')
+			deliver()
+		})
+
+		this.on('close', () => {
+			this.log('close :: connected set to `false`')
+			this.connected = false
+
+			this.log('close :: clearing connackTimer')
+			clearTimeout(this.connackTimer)
+
+			this.log('close :: clearing ping timer')
+			if (this.pingTimer !== null) {
+				this.pingTimer.clear()
+				this.pingTimer = null
+			}
+
+			if (this.topicAliasRecv) {
+				this.topicAliasRecv.clear()
+			}
+
+			this.log('close :: calling _setupReconnect')
+			this._setupReconnect()
+		})
+
+		if (!this.options.manualConnect) {
+			this.log('MqttClient :: setting up stream')
+			this.connect()
+		}
+	}
+
+	/**
+	 * Setup the event handlers in the inner stream, sends `connect` and `auth` packets
+	 */
+	connect() {
+		const writable = new Writable()
+		const parser = mqttPacket.parser(this.options)
+		let completeParse = null
+		const packets = []
+
+		this.log('connect :: calling method to clear reconnect')
+		this._clearReconnect()
+
+		this.log(
+			'connect :: using streamBuilder provided to client to create stream',
+		)
+		this.stream = this.streamBuilder(this)
+
+		parser.on('packet', (packet) => {
+			this.log('parser :: on packet push to packets array.')
+			packets.push(packet)
+		})
+
+		const work = () => {
+			this.log('work :: getting next packet in queue')
+			const packet = packets.shift()
+
+			if (packet) {
+				this.log('work :: packet pulled from queue')
+				handlePacket(this, packet, nextTickWork)
+			} else {
+				this.log('work :: no packets in queue')
+				const done = completeParse
+				completeParse = null
+				this.log('work :: done flag is %s', !!done)
+				if (done) done()
+			}
+		}
+
+		const nextTickWork = () => {
+			if (packets.length) {
+				nextTick(work)
+			} else {
+				const done = completeParse
+				completeParse = null
+				done()
+			}
+		}
+
+		writable._write = (buf, enc, done) => {
+			completeParse = done
+			this.log('writable stream :: parsing buffer')
+			parser.parse(buf)
+			work()
+		}
+
+		const streamErrorHandler = (error) => {
+			this.log('streamErrorHandler :: error', error.message)
+			if (socketErrors.includes(error.code)) {
+				// handle error
+				this.log('streamErrorHandler :: emitting error')
+				this.emit('error', error)
+			} else {
+				this.nop(error)
```

Comment:
nit: I think changing the name here would make things slightly more clear. Not sure if this was a typo that carried on for a long time or if it was intentional. 
```suggestion
				this.noop(error)
```

Reply:
Done